### DISCUSSION TITLE
Get rid of "unready" strings.

### DISCUSSION THREAD

**Discussion by markshannon** [faster-cpython/ideas#156] [General]
The CPython codebase is riddled* with `PyUnicode_READY` checks. Having to check this on the fast path of comparisons and other operations on strings is highly wasteful.
Not only that, but it is hard to be sure that we are sufficiently defensive and that we call `PyUnicode_READY` in all places where we need to.

Strings are immutable at the Python level, they should also be immutable at the C level.

The `PyUnicode_READY` seems to exist only to provide a false convenience when writing C code that creates strings.

I don't know what the best approach is. I suspect that a C-API change will be needed, but maybe there are some other possibilities.

One idea I have is to not create a `str`, but some sort of `UninitializedStr` instance in the offending C-API.
The `UninitializedStr` should then attempt to convert itself to a `str` when involved in any operation.
Not sure how well that would work when the `UninitializedStr` is passed to an APIs that expect a `str`.

@methane, as the author of PEP 623, do you have any ideas or suggestions?

\* About 260 checks.

**Comment by methane** [comment]
PEP 623 has been accepted and PyUnicode_READY() will be removed (or become no-op macro) in Python 3.12.

**Comment by markshannon** [comment]
Fantastic!
